### PR TITLE
ARROW-10304: [C++][Compute] Optimize variance kernel for integers

### PR DESCRIPTION
**Pull request by cyb70289**
Improve variance kernel performance for integers by leveraging
textbook one pass algorithm and integer arithmetic.

### CODE REVIEW

**Comment by cyb70289** [comment]
**NOTE:** Benchmark PR https://github.com/apache/arrow/pull/8407 is not merged yet. Need to manually pull that PR to evaluation performance.

Tested on Xeon Gold 5218, clang-9.

```
                             benchmark         baseline        contender  change %
8        VarianceKernelInt32/1048576/0    1.798 GiB/sec    6.742 GiB/sec   274.842   'null_percent': 0.0
18   VarianceKernelInt32/1048576/10000    1.742 GiB/sec    5.155 GiB/sec   195.990  'null_percent': 0.01
6      VarianceKernelInt32/1048576/100    1.135 GiB/sec    2.806 GiB/sec   147.285   'null_percent': 1.0
7       VarianceKernelInt32/1048576/10  858.921 MiB/sec    1.551 GiB/sec    84.901  'null_percent': 10.0
19       VarianceKernelInt32/1048576/2  456.209 MiB/sec  771.633 MiB/sec    69.140  'null_percent': 50.0
12       VarianceKernelFloat/1048576/2  396.262 MiB/sec  455.307 MiB/sec    14.900  'null_percent': 50.0
14       VarianceKernelFloat/1048576/1    1.095 TiB/sec    1.210 TiB/sec    10.563  null_percent': 100.0
3       VarianceKernelFloat/1048576/10  788.471 MiB/sec  857.623 MiB/sec     8.770  'null_percent': 10.0
15      VarianceKernelDouble/1048576/1    1.152 TiB/sec    1.213 TiB/sec     5.312  null_percent': 100.0
1        VarianceKernelInt32/1048576/1    1.200 TiB/sec    1.259 TiB/sec     4.901  null_percent': 100.0
23     VarianceKernelFloat/1048576/100    1.129 GiB/sec    1.134 GiB/sec     0.371   'null_percent': 1.0
4       VarianceKernelDouble/1048576/0    3.580 GiB/sec    3.587 GiB/sec     0.195   'null_percent': 0.0
20       VarianceKernelFloat/1048576/0    1.798 GiB/sec    1.800 GiB/sec     0.090   'null_percent': 0.0
10   VarianceKernelFloat/1048576/10000    1.743 GiB/sec    1.743 GiB/sec     0.006  'null_percent': 0.01
2    VarianceKernelInt64/1048576/10000    3.468 GiB/sec    3.468 GiB/sec    -0.005  'null_percent': 0.01
17       VarianceKernelInt64/1048576/0    3.586 GiB/sec    3.584 GiB/sec    -0.050   'null_percent': 0.0
0   VarianceKernelDouble/1048576/10000    3.476 GiB/sec    3.474 GiB/sec    -0.066  'null_percent': 0.01
11    VarianceKernelDouble/1048576/100    2.276 GiB/sec    2.274 GiB/sec    -0.122   'null_percent': 1.0
5      VarianceKernelInt64/1048576/100    2.274 GiB/sec    2.269 GiB/sec    -0.227   'null_percent': 1.0
22     VarianceKernelDouble/1048576/10    1.652 GiB/sec    1.604 GiB/sec    -2.856  'null_percent': 10.0
21       VarianceKernelInt64/1048576/1    1.242 TiB/sec    1.206 TiB/sec    -2.919  null_percent': 100.0
13      VarianceKernelDouble/1048576/2  864.392 MiB/sec  832.143 MiB/sec    -3.731  'null_percent': 50.0
16      VarianceKernelInt64/1048576/10    1.681 GiB/sec    1.546 GiB/sec    -8.006  'null_percent': 10.0
9        VarianceKernelInt64/1048576/2  910.548 MiB/sec  786.507 MiB/sec   -13.623  'null_percent': 50.0
```

**Comment by cyb70289** [comment]
Will add 64bit integers optimization.

**Comment by cyb70289** [comment]
Added int64 optimization. Updated benchmark result. Ready for review.
Big improvement for `int32`. Moderate improvement for `int64` with few null values.
Some drop for `int64` with many null values.

File: cpp/src/arrow/compute/kernels/aggregate_var_std.cc [hunk 1]
```diff
@@ -57,6 +63,54 @@ struct VarStdState {
...
+              square_sum += static_cast<uint64_t>(value) * value;
+            },
+            []() {});
+
+        const double mean = static_cast<double>(sum) / count;
+        // calculate m2 = square_sum - sum * sum / count
+        // decompose `sum * sum / count` into integers and fractions
+        const int128_t sum_square = static_cast<int128_t>(sum) * sum;
+        const int128_t integers = sum_square / count;
+        const double fractions = static_cast<double>(sum_square % count) / count;
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_var_std.cc:101]: Or simply `sum_square - integers * count`?

Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_var_std.cc:101]: `fractions` will be a floating number less than 1. Do you mean changing this line to
`const double fractions = static_cast<double>(sum_square - integers * count) / count;` ?

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/aggregate_var_std.cc:101]: Hmm, yes.

Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_var_std.cc:101]: Looks compiler is able to figure out the pattern and generate same code, without unnecessary division.
https://godbolt.org/z/dWY9Ex

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/aggregate_var_std.cc:101]: Great, thank you.

**Comment by pitrou** [comment]
Results on an AMD Zen 2 CPU:
```
VarianceKernelInt32/1048576/10000         140 us          140 us         5030 bytes_per_second=6.98658G/s null_percent=0.01 size=1048.58k
VarianceKernelInt32/1048576/100           216 us          216 us         3267 bytes_per_second=4.5294G/s null_percent=1 size=1048.58k
VarianceKernelInt32/1048576/10            397 us          397 us         1763 bytes_per_second=2.45765G/s null_percent=10 size=1048.58k
VarianceKernelInt32/1048576/2             974 us          974 us          718 bytes_per_second=1026.87M/s null_percent=50 size=1048.58k
VarianceKernelInt32/1048576/1           0.816 us        0.816 us       844145 bytes_per_second=1.1684T/s null_percent=100 size=1048.58k
VarianceKernelInt32/1048576/0             130 us          130 us         5414 bytes_per_second=7.51569G/s null_percent=0 size=1048.58k

VarianceKernelInt64/1048576/10000         135 us          135 us         5174 bytes_per_second=7.22877G/s null_percent=0.01 size=1048.58k
VarianceKernelInt64/1048576/100           260 us          260 us         2682 bytes_per_second=3.7503G/s null_percent=1 size=1048.58k
VarianceKernelInt64/1048576/10            440 us          440 us         1591 bytes_per_second=2.21931G/s null_percent=10 size=1048.58k
VarianceKernelInt64/1048576/2             884 us          884 us          783 bytes_per_second=1.10507G/s null_percent=50 size=1048.58k
VarianceKernelInt64/1048576/1           0.821 us        0.821 us       840316 bytes_per_second=1.16182T/s null_percent=100 size=1048.58k
VarianceKernelInt64/1048576/0             123 us          123 us         5620 bytes_per_second=7.94262G/s null_percent=0 size=1048.58k

VarianceKernelFloat/1048576/10000         366 us          366 us         1909 bytes_per_second=2.66576G/s null_percent=0.01 size=1048.58k
VarianceKernelFloat/1048576/100           751 us          751 us          909 bytes_per_second=1.3003G/s null_percent=1 size=1048.58k
VarianceKernelFloat/1048576/10           1097 us         1097 us          637 bytes_per_second=911.712M/s null_percent=10 size=1048.58k
VarianceKernelFloat/1048576/2            1803 us         1802 us          387 bytes_per_second=554.854M/s null_percent=50 size=1048.58k
VarianceKernelFloat/1048576/1           0.817 us        0.817 us       838993 bytes_per_second=1.1679T/s null_percent=100 size=1048.58k
VarianceKernelFloat/1048576/0             346 us          346 us         2021 bytes_per_second=2.82409G/s null_percent=0 size=1048.58k

VarianceKernelDouble/1048576/10000        184 us          184 us         3751 bytes_per_second=5.30153G/s null_percent=0.01 size=1048.58k
VarianceKernelDouble/1048576/100          372 us          372 us         1869 bytes_per_second=2.62218G/s null_percent=1 size=1048.58k
VarianceKernelDouble/1048576/10           549 us          549 us         1249 bytes_per_second=1.77993G/s null_percent=10 size=1048.58k
VarianceKernelDouble/1048576/2            909 us          909 us          741 bytes_per_second=1099.92M/s null_percent=50 size=1048.58k
VarianceKernelDouble/1048576/1          0.831 us        0.831 us       831173 bytes_per_second=1.14779T/s null_percent=100 size=1048.58k
VarianceKernelDouble/1048576/0            174 us          174 us         4050 bytes_per_second=5.62431G/s null_percent=0 size=1048.58k
```

I'm curious why Int64 would be faster than Double. Aren't they using the same algorithm? (and Int64 goes through an additional int-to-float conversion for each value)

**Comment by cyb70289** [comment]
> I'm curious why Int64 would be faster than Double. Aren't they using the same algorithm? (and Int64 goes through an additional int-to-float conversion for each value)

There's no int-to-float conversion in Int64 summation loop (sum to Int128). It's faster than double summation.
https://quick-bench.com/q/-P9E6tgtXqnVBpVmmN6piaZHeUA

**Comment by pitrou** [comment]
Ah, I hadn't noticed the `SumType`.

**Review by pitrou — APPROVED** [review]
+1

**Pull request closed without merging**